Add functions to spatially visualise individual and aggregate criteria for location selection

Rosejoycrocker opened the pull request:
Closes #620

Adds `viz.decision_matrices` to visualise specified criteria and aggregate criteria spatially from a decision matrix, aggregate score vector and result set.

Adds `analysis.decision_matrices` which calculates decision matrices for seeding and fogging from a result set and outputs these as `NamedDimsArrays` labelled using the criteria weightings names.

Some changes to criteria weights names have also been added for consistency and ease of labelling the decision matrices as `NamedDimsArray`. These changes will make addressing #624 easier. 

The weighting for ingoing and outgoing connectivity for fogging has been split into two separate weightings `fog_in_connectivity` and `fog_out_connectivity` for consistency with seeding.  The weights `coral_cover_low` and `coral_cover_high` have been renamed `seed_coral_cover_low` and `fog_coral_cover_high` for consistency with other weights names. The weights ordering in `CriteriaWeights` has been reordered so that `criteria_params` extracts the weights names in the correct order to label the decision matrices.

Zapiano commented:
@Rosejoycrocker when I save the file `ext\AvizExt\viz\location_selection.jl`, Julia Formatter changes a lot of things. Is your Julia Formatter working?

Rosejoycrocker commented:
> @Rosejoycrocker when I save the file `ext\AvizExt\viz\location_selection.jl`, Julia Formatter changes a lot of things. Is your Julia Formatter working?

Hi @Zapiano, I recently changed laptops and had to set up my formatter again. I thought I finally had it working but obviously not, I'll look into it again.

Zapiano reviewed (commented):
I did an initial review but it's too hard to review without the autoformatted fixed....

Rosejoycrocker commented on ext/AvizExt/viz/location_selection.jl:
```diff
@@ -102,32 +104,134 @@ function ADRIA.viz.ranks_to_frequencies!(
+# Arguments
+- `g` : Figure GridPosition or GridLayout
+- `rs` : Result set
+- `S` : A normalised decision matrix calculated using decison.decision_matrices
```
`SE` and `SH` are used for the seeding and shading matrices respectively, but I'll change it to `decision_matrix` for clarity

Rosejoycrocker commented on ext/AvizExt/viz/location_selection.jl:
```diff
@@ -102,32 +104,134 @@ function ADRIA.viz.ranks_to_frequencies!(
+    axis_opts::Dict = Dict(),
+)
+    if length(rs.site_data.site_id) != size(S, 1)
+        error("Only unfiltered decision matrices can be plotted.")
```
The decision matrices can be filtered based on thresholds for different criteria (like heat stress and wave stress) but this means they have less sites than the whole site set, which means when you go to plot them the geodata (created from the result set) doesn't match the number of sites when plotting using `map`. Alternatively, filtered sites could be set as `0.0` or something, but they'd have to be added back into the decision matrix because these are automatically filtered in `create_decision_matrix`.

Rosejoycrocker commented on ext/AvizExt/viz/location_selection.jl:
```diff
@@ -102,32 +104,134 @@ function ADRIA.viz.ranks_to_frequencies!(
+    try
+        # Clear empty figures
+        trim!(g)
+    catch err
```
Good point, this is to catch an error when only one figure is plotted, but because of the score+ at least 1 criteria you would never have this occur

Rosejoycrocker commented on src/analysis/decision.jl:
```diff
@@ -0,0 +1,109 @@
+# Arguments
+- `rs` : ADRIA ResultSet
+- `criteria_weights` :  A row of a scenario dataframe, containing intervention criteria weights.
+- `int_type` : Intervention type (e.g. :seed or :fog)
```
This is no longer an input, I'll remove it from docs, thanks for catching :)

Rosejoycrocker commented on src/analysis/decision.jl:
```diff
@@ -0,0 +1,109 @@
+    RCP::String = "45",
+)
+    site_ids = collect(1:length(rs.site_data.site_id))
+    leftover_space = relative_leftover_space(loc_coral_cover) .* site_k_area(rs)
```
This is how `leftover_space` is calculated in `scenario.jl` - should I add a function for this @ConnectedSystems, or was there a reason for setting it up this way?

Rosejoycrocker commented on src/analysis/analysis.jl:
```diff
@@ -93,15 +101,16 @@ Computes confidence interval for series of data.
 include("clustering.jl")
 include("rule_extraction.jl")
 include("scenario.jl")
+include("decision.jl")
```
Mostly because the function uses the result set so I thought it was better suited to analysis, but you're right it's not ideal. What do you think @ConnectedSystems ?

Rosejoycrocker commented on src/analysis/decision.jl:
```diff
@@ -0,0 +1,109 @@
+        rs.dhw_stats[RCP](; stat = "mean"), (:dhw_scenario)
+    )
+
+    TP_data = rs.connectivity_data[RCP]  # connectivity matrix
```
`TP_base` is what the connectivity matrix is called in the connectivity calcs: https://github.com/open-AIMS/ADRIA.jl/blob/283084030f360f7de4a3db0f06879d47d14b78ca/src/ecosystem/connectivity.jl#L171-L176
but you're right it's not a very transparent name. I'll change it to `conn_data` which is also used in other places in ADRIA.

Rosejoycrocker commented on src/analysis/decision.jl:
```diff
@@ -0,0 +1,109 @@
+
+    # Criteria for strongest larval sources to priority locations
+    priority_source_criteria = ADRIA.decision.priority_predecessor_criteria(
+        strong_pred, vec(rs.sim_constants["priority_sites"]), length(site_ids)
```
It's only used once, the other usages is `vec(rs.sim_constants["priority_zones"])`

Rosejoycrocker commented on src/analysis/decision.jl:
```diff
@@ -0,0 +1,109 @@
+
+    TP_data = rs.connectivity_data[RCP]  # connectivity matrix
+    connectivity_data = connectivity_strength(
+        TP_data .* site_k_area(rs), vec(loc_coral_cover), TP_data
```
Good point, not sure why this is here

ConnectedSystems commented on src/ecosystem/connectivity.jl:
```diff
@@ -179,47 +179,49 @@ NamedTuple:
-
-    return (in_conn=C1, out_conn=C2, strongest_predecessor=strong_pred)
+function connectivity_strength(
+	TP_base::AbstractMatrix{<:Union{Float32, Float64}}
```
I recommend keeping the union: allows the compiler to specialise for just two types (Float32 and Float64) compared to having to support all possible Reals

Rosejoycrocker commented:
> > @Rosejoycrocker when I save the file `ext\AvizExt\viz\location_selection.jl`, Julia Formatter changes a lot of things. Is your Julia Formatter working?
> 
> Hi @Zapiano, I recently changed laptops and had to set up my formatter again. I thought I finally had it working but obviously not, I'll look into it again.

Hi @Zapiano, I think I have the formatter working now (ended up doing a full uninstall and reinstall of VS code), but let me know if some things look odd. I've addressed your comments and also changed some things suggested by @ConnectedSystems, such as making `viz.decision_matrices` a method of `viz.map` .

ConnectedSystems commented:
Bump! Where did we end up on this @Zapiano ?

Rosejoycrocker commented:
> @Rosejoycrocker this needs a rebase before I review again. I'm so sorry it took me so long...
> 
> I strongly suggest you don't simply rebase it though, because
> 
> 1. the example csv was changed so you better generate a new one.
> 2. there seem to be a lot of formatting changes and recently Arlo updated the formatting for the entire codebase
> 3. you have almost 40 commits but what you actually want to change seems to be way less than what this is actually changing.
> 
> I think it might be more productive and avoid mistakes to just create a `tmp-mcda-visualisations` branch from main, manually bring the relevant changes to that and reset this branch to the tmp one. 😢 (I know it is annoying but I honestly think this will be faster and easier to do and that's what I would do probably).
> 
> Let me know if you have any questions (or if you hate me now)

@Zapiano honestly I forgot about this because it's been 8 months since I responded to your review and naturally much has changed in Main since then. I agree the best approach is probably to start a new branch although I won't be able to give time to this until October. Of course I don't hate you!
